City-of-Helsinki/hel-fi-drupal-grants
# feature: AU-2148, E2E, Made profile tests more robust

# [AU-2148](https://helsinkisolutionoffice.atlassian.net/browse/AU-2148)

## What was done

This pull request attempts to fix the issue where application tests would sometimes redirect to the profile page because of a missing profile or incomplete profile data. The issue has been fixed by implementing the following changes:

1. **New profile validation**: When the tests are executed and a new profile is created either because the existing profile is over one (1) hour old, the profile has missing data, or because profile creation is force through the `CREATE_PROFILE` .env variable, we now validate the created profile and its data. The validation is performed in the same manner as application data validation: We visit the user profile page and validate that the submitted profile data is present on the profile page. This change has been implemented for all profile roles.

2. **Existing profile validation**: When the tests are executed and we don't create a new profile (either because the profile isn't old enough, the profile does not have missing data, or the creation is not forced), we still valid the existing profile and its data. This is done my visiting the profile page and making sure that the hard-coded `PROFILE_INPUT_DATA` data from `profile_input_data.ts` is present on the profile page. This change has been implemented for all profile roles.

## What has changed

- Refactored and implemented logic changes to all the profile tests files: `registered_community.ts`, `unregistered_community.ts` and `private_person.ts`. The files now contain the follow tests:

  - `Profile form tests`: Runs all the form variant tests for a given profile type. This only runs if the profiles is over one (1) hours old, if the profile has missing data, or if the creation is forced through an .env variable.
  
  - `Validate profile data`: Validates the submitted profile data on the profile page (just the success test for now). This only runs if the profiles is over one (1) hours old, if the profile has missing data, or if the creation is forced through an .env variable.
  
  - `Validate existing profile`: Validates an existing profiles data. This only runs if the two pervious tests are skipped.

- Cleaned up and added necessary validation keys to all the profile test data files in `profile_data_registered_community.ts`, `profile_data_unregistered_community.ts` and `profile_data_private_person.ts`: The formatting and indentation was wrong in all of the files so I fixed them. I also added the necessary `viewPageSelector` and `viewPageSkipValidation` keys so that validation can be performed.

## Other smaller changes

- `document_helpers.ts`: 
  - Cleaned up code.
  - Fixed code so that TypeScript doesn't show errors.

- `form_helpers.ts`: 
  - Added a check for making sure a profile exists before filling an application form.
  - Refactored the `uploadFile()` function so that is uses the passed in result value. Removed hard-coded wait times which were causing problems when the dev environment was running slow.
  - Adjusted the error handling in `fillProfileForm()` so that messages are not logged unnecessarily.
  - Cleaned up code.

- `profile_helpers.ts`: 
  - Implemented the `runProfileFormTest()` function that's utilized in the `Profile form tests` test.
  - Added a check to `fetchLatestProfileByType()` for missing profile content.
  - Cleaned up code. 
  - Removed deprecated code. 

- `validation_helpers.ts`: 
  - Added profile data validation. Implemented the functions `validateProfileData()` and `validateExistingProfileData()` for both dynamic and existing profile data validation.

- `profile_input_data.ts`: 
  - Added more hard-coded data to the profile so that the existing profile check is thorough.

- `application_data_60ts`: 
  - Fixed a test with wrong values.

- `application_data_61ts`: 
  - Added missing `viewPageSkipValidation` keys to `missingValues` and `wrongValues` form variants.

- `registered_community_61.ts`: 
  - Fixed the validation test implementation.

- `application_data_61ts`: 
  - Fixed a test with a typo.

- `playwright.config.ts`: 
  - Fixed a typo. Added a few missing tests.

## How to install

Make sure your instance is up and running on correct branch.
* `git checkout feature/AU-2148-e2e-missing-profile-problem`
* `make fresh`
* `make drush-cr`

## How to test
- Set `CREATE_PROFILE="TRUE"` in your `.env` file. Run `make test-pw-p PROJECT=forms-48`. Make sure both the `Profile form tests` and `Validate profile data` tests are **run** and `Validate existing profile` is **skipped** for each role.

![Screenshot 2024-03-27 at 14 26 03](https://github.com/City-of-Helsinki/hel-fi-drupal-grants/assets/117807015/d01d866c-9f5b-4cf5-8027-22d340d9869c)

- Set `CREATE_PROFILE="FALSE"` in your `.env` file. Run `make test-pw-p PROJECT=forms-48`. Make sure both the `Profile form tests` and `Validate profile data` tests are **skipped** and `Validate existing profile` is **run** for each role.

![Screenshot 2024-03-27 at 14 27 43](https://github.com/City-of-Helsinki/hel-fi-drupal-grants/assets/117807015/1da2cda5-621b-47de-9af6-9d9180a83c70)

Now you are going to attempt to cause an "error state" which should be caught by the new tests.
- Set `CREATE_PROFILE="TRUE"` in your `.env` file and run `make test-pw-p PROJECT=forms-48`. Interrupt the tests by hitting `CTRL + C` as soon as you see the line `Deleted 1 grant profiles from ATV` in the terminal. This will cause a state where ATV has no profile data, but the profile has been updated during the last hour.

![Screenshot 2024-03-27 at 14 29 37](https://github.com/City-of-Helsinki/hel-fi-drupal-grants/assets/117807015/4b5632c7-d224-4ba8-945d-aead0a858517)

- Now, set `CREATE_PROFILE="FALSE"` in your `.env` file and run `make test-pw-p PROJECT=forms-48` again. You should be met with the message `...has missing content. Re-creating.`, and both the `Profile form tests` and `Validate profile data` test should run normally.

![Screenshot 2024-03-27 at 14 31 08](https://github.com/City-of-Helsinki/hel-fi-drupal-grants/assets/117807015/bdc76845-ff35-43d2-a88b-1b6210b2eb8a)


- If you have time, run `make test-pw` to make sure all the tests pass.

[AU-2148]: https://helsinkisolutionoffice.atlassian.net/browse/AU-2148?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
